src/internal/serviceenv/service_env.go

```diff
@@ -172,7 +173,12 @@ func InitServiceEnv(config *Configuration) *NonblockingServiceEnv {
 		env.dbEg.Go(env.initDirectDBClient)
 	}
 	env.listener = env.newListener()
-	if lokiHost, lokiPort := os.Getenv(env.config.LokiHostVar), os.Getenv(env.config.LokiPortVar); lokiHost != "" && lokiPort != "" {
+	if lokiHost, lokiPort := env.config.LokiHost, env.config.LokiPort; lokiHost != "" && lokiPort != "" {
+		env.lokiClient = &loki.Client{
+			Address: fmt.Sprintf("http://%s", net.JoinHostPort(lokiHost, lokiPort)),
+		}
+	} else if lokiHost, lokiPort := os.Getenv(env.config.LokiHostVar), os.Getenv(env.config.LokiPortVar); lokiHost != "" && lokiPort != "" {
```

Comment:
It turns out that I don't know why we bother doing this environment variable evaluation in pachd.  It seems that environment variables in a k8s manifest can refer to existing environment variables, so we can make LOKI_SERVICE_HOST (etc.) the only configuration we take, and then have helm generate:

```yaml
env:
  - name: LOKI_SERVICE_HOST
    value:  {{ if .lokiServiceHost }} {{ .lokiServiceHost }} {{ else }} $({{ release.Name | ... }}_LOKI_SERVICE_HOST)
```

So that will evaluate to a value like `literal-hostname.whatever` or `$(FOOBAR_LOKI_SERVICE_HOST)` depending on external loki or built-in loki.  Then pachd doesn't have to do this work.

Of course, we should check and see if that actually works.  The service environment variables we're depending on to exist are pretty undocumented.  (I read the code in the kubelet, though, and don't see why it wouldn't work.)

Reply:
Oh this is way better. The variable indirection seemed unnecessary but I figured maybe it was in case the service IP or port changes during runtime. That said, if it changes during runtime, it doesn't matter since we only set the lokiClient one time anyways.

Reply:
It's possible that it doesn't work, or whoever added this feature didn't know k8s could substitute environment variables references.  The docs say it can, but I've never tried it, and they don't say what syntax to use.  I just guessed 😂

Reply:
It works!

Helm deployment.yaml
```yaml
        - name: LOKI_SERVICE_HOST
        {{- if .Values.pachd.lokiServiceHost }}
          value:  {{ .Values.pachd.lokiServiceHost }}
        {{ else }}
          value: $({{ snakecase .Release.Name | upper }}_LOKI_SERVICE_HOST)
        {{- end }}
```
```bash
kubectl exec -it pachd-699bd4cd9d-wmb5f -- env | grep LOKI
LOKI_SERVICE_PORT=3100
LOKI_SERVICE_HOST=10.100.47.53
LOKI_SERVICE_HOST_VAR=PACHYDERM_LOKI_SERVICE_HOST
LOKI_SERVICE_PORT_VAR=PACHYDERM_LOKI_SERVICE_PORT
LOKI_LOGGING=true
PACHYDERM_LOKI_PORT_3100_TCP_ADDR=10.100.47.53
PACHYDERM_LOKI_PORT_3100_TCP_PROTO=tcp
PACHYDERM_LOKI_PORT=tcp://10.100.47.53:3100
PACHYDERM_LOKI_SERVICE_PORT_HTTP_METRICS=3100
PACHYDERM_LOKI_SERVICE_HOST=10.100.47.53
PACHYDERM_LOKI_PORT_3100_TCP_PORT=3100
PACHYDERM_LOKI_SERVICE_PORT=3100
PACHYDERM_LOKI_PORT_3100_TCP=tcp://10.100.47.53:3100
```